pkg/gatewayconfigurationserver/http.go

```diff
@@ -0,0 +1,117 @@
+// Copyright © 2022 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gatewayconfigurationserver
+
+import (
+	"context"
+	"encoding"
+	"encoding/json"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"go.thethings.network/lorawan-stack/v3/pkg/pfconfig/cpf"
+	"go.thethings.network/lorawan-stack/v3/pkg/pfconfig/semtechudp"
+	"go.thethings.network/lorawan-stack/v3/pkg/ratelimit"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/web"
+	"go.thethings.network/lorawan-stack/v3/pkg/webhandlers"
+	"go.thethings.network/lorawan-stack/v3/pkg/webmiddleware"
+)
+
+// RegisterRoutes registers the web frontend routes.
+func (s *Server) RegisterRoutes(server *web.Server) {
```

Comment:
Aren't we replacing these routes with the gRPC service? 

The gRPC gateway paths are defined in your new proto as 
```
"/gcs/gateways/{gateway_ids.gateway_id}/configuration/{format}"
```

If you want a Semtech UDP file from this web route, you'd make a request like

1. Through gRPC gateway
```
/gcs/gateways/{gateway_ids.gateway_id}/configuration/semtechudp
```

2. Through this web route
```
/gcs/gateways/{gateway_id}/semtechudp/global_conf.json
```

These look like they do the same thing. 

If you add a `string filename` field to your `GetGatewayConfigurationRequest` message and remove the "configuration" word, then the gRPC request would become
```
/gcs/gateways/{gateway_ids.gateway_id}/{format}/{filename}
```

This is the same as what's duplicated in these web routes right? So why not make harmonize these and get rid of the web routes?

Reply:
I see it, the http and grpc routes could be merged. In addition of the `filename` field could open space for more configuration files without adding extra routes. 

Will work on that change.

Reply:
It slipped by me when looking it at first  but there is a problem with this solution, it becomes more clear when looking at the previously existent paths:
- `/gcs/gateways/{gateway_id}/semtechudp/global_conf.json`
- `/gcs/gateways/{gateway_id}/kerlink-cpf/lorad/lorad.json`
- `/gcs/gateways/{gateway_id}/kerlink-cpf/lorafwd/lorafwd.toml`

The suggestion `/gcs/gateways/{gateway_ids.gateway_id}/{format}/{filename}` becomes something where `{format}` has to include something like `kerlink-cpf/loradfwd` which from my tests locally doesn't seem possible to do.

I'm a fan of making a gRPC gateway that replaces the the three previously defined http routes but I'm struggling to see a solution that doesn't change the api. So in my mind it might be more worth it to leave it as is.

Reply:
@KrishnaIyer small reminder of this discussion

Reply:
Ok in that case can you add a `type` parameter? 
Then you can have `/gcs/gateways/{gateway_ids.gateway_id}/{format}/{type}/{filename}`.

For the Semtech UDP, since the type is not needed, you can add an [additional binding](https://github.com/TheThingsNetwork/lorawan-stack/blob/1c018c71892ef2c258b99fc751ca44efa1abb8a3/api/application_services.proto#L165) to that path without the type.

Reply:
From our discussion and tests, the problem with replacing the existent routes with the ones that are generated by the gRPC-gateway is the fact that the response body changes in such a way that the json of the configuration stays as the base64 of the byte array. So when doing a request to the GCS gRPC service the response will alway be:
```json
{
   "contents": "<base64_byte_array>"
}
```

Since this changes the API, is best to keep the old implementation.